File: src/LTAdmin/Services/Auth/Habilitations.cs
using LTAdmin.Data;

namespace LTAdmin.Services.Auth;

/// <summary>
/// Modules visibles dans le menu (libellés métier, jamais de noms de tables)
/// et habilitations par profil.
/// </summary>
public static class Modules
{
    public const string TableauDeBord = "Tableau de bord";
    public const string Etudiants = "Étudiants";
    public const string Inscriptions = "Inscriptions";
    public const string Referentiel = "Référentiel";
    public const string Formateurs = "Formateurs";
    public const string Notes = "Notes et évaluations";
    public const string Bulletins = "Bulletins";
    public const string Examens = "Examens";
    public const string EmploiDuTemps = "Emploi du temps";
    public const string Absences = "Absences";
    public const string Ecolage = "Écolage";
    public const string Paie = "Paie des formateurs";
    public const string Statistiques = "Statistiques";
    public const string Rapports = "Rapports";
    public const string Administration = "Administration";
    public const string Utilisateurs = "Profils utilisateurs";
}

public static class Habilitations
{
    public static readonly string[] MenuOrder =
    {
        Modules.TableauDeBord,
        Modules.Etudiants,
        Modules.Inscriptions,
        Modules.Referentiel,
        Modules.Formateurs,
        Modules.Notes,
        Modules.Bulletins,
        Modules.Examens,
        Modules.EmploiDuTemps,
        Modules.Absences,
        Modules.Ecolage,
        Modules.Paie,
        Modules.Statistiques,
        Modules.Rapports,
        Modules.Administration
    };

    public static string NormalizeProfil(string? profil)
    {
        var p = (profil ?? "").Trim();
        if (p.Equals(Profils.Admin, StringComparison.OrdinalIgnoreCase)
            || p.Equals(Profils.Administrateur, StringComparison.OrdinalIgnoreCase)
            || p.Equals("Administrateur", StringComparison.OrdinalIgnoreCase))
            return Profils.Administrateur;
        if (p.Equals(Profils.Direction, StringComparison.OrdinalIgnoreCase)
            || p.Equals("DIRECTION", StringComparison.OrdinalIgnoreCase))
            return Profils.Direction;
        if (p.Equals(Profils.Scolarite, StringComparison.OrdinalIgnoreCase)
            || p.Equals(Profils.ScolariteLibelle, StringComparison.OrdinalIgnoreCase)
            || p.Equals("SCOLARITE", StringComparison.OrdinalIgnoreCase))
            return Profils.ScolariteLibelle;
        if (p.Equals(Profils.Finance, StringComparison.OrdinalIgnoreCase)
            || p.Equals(Profils.Comptabilite, StringComparison.OrdinalIgnoreCase)
            || p.Equals("COMPTABILITE", StringComparison.OrdinalIgnoreCase)
            || p.Equals("CAISSE", StringComparison.OrdinalIgnoreCase))
            return Profils.Comptabilite;
        if (p.Equals(Profils.Enseignant, StringComparison.OrdinalIgnoreCase)
            || p.Equals("ENSEIGNANT", StringComparison.OrdinalIgnoreCase)
            || p.Equals("FORMATEUR", StringComparison.OrdinalIgnoreCase))
            return Profils.Enseignant;
        return string.IsNullOrWhiteSpace(p) ? "Utilisateur" : p;
    }

    public static bool CanAccess(string? profil, string module)
    {
        if (string.Equals(module, Modules.TableauDeBord, StringComparison.OrdinalIgnoreCase))
            return true;

        var role = NormalizeProfil(profil);
        if (role == Profils.Administrateur)
            return true;

        return role switch
        {
            var r when r == Profils.Direction => module is
                Modules.Etudiants or Modules.Inscriptions or Modules.Referentiel
                or Modules.Formateurs or Modules.Notes or Modules.Bulletins
                or Modules.Examens or Modules.EmploiDuTemps or Modules.Absences
                or Modules.Ecolage or Modules.Paie or Modules.Statistiques
                or Modules.Rapports,
            var r when r == Profils.ScolariteLibelle => module is
                Modules.Etudiants or Modules.Inscriptions or Modules.Referentiel
                or Modules.Formateurs or Modules.Notes or Modules.Bulletins
                or Modules.Examens or Modules.EmploiDuTemps or Modules.Absences
                or Modules.Statistiques or Modules.Rapports,
            var r when r == Profils.Comptabilite => module is
                Modules.Ecolage or Modules.Paie or Modules.Statistiques or Modules.Rapports,
            var r when r == Profils.Enseignant => module is
                Modules.Notes or Modules.Bulletins or Modules.Examens
                or Modules.EmploiDuTemps or Modules.Absences or Modules.Formateurs,
            _ => false
        };
    }

    public static IReadOnlyList<string> ProfilsConnus { get; } = new[]
    {
        Profils.Administrateur,
        Profils.Direction,
        Profils.ScolariteLibelle,
        Profils.Comptabilite,
        Profils.Enseignant
    };
}

File: ltadmin/services/auth/habilitations.py
"""Modules du menu et habilitations par profil (portage de Habilitations.cs)."""

from __future__ import annotations

from typing import List, Optional

from ltadmin.data.schema import Profils


class Modules:
    """Libellés métier du menu (jamais de noms de tables)."""

    TABLEAU_DE_BORD = "Tableau de bord"
    ETUDIANTS = "Étudiants"
    INSCRIPTIONS = "Inscriptions"
    REFERENTIEL = "Référentiel"
    FORMATEURS = "Formateurs"
    NOTES = "Notes et évaluations"
    BULLETINS = "Bulletins"
    EXAMENS = "Examens"
    EMPLOI_DU_TEMPS = "Emploi du temps"
    ABSENCES = "Absences"
    ECOLAGE = "Écolage"
    PAIE = "Paie des formateurs"
    STATISTIQUES = "Statistiques"
    RAPPORTS = "Rapports"
    ADMINISTRATION = "Administration"
    UTILISATEURS = "Profils utilisateurs"
    TABLES = "Tables (maintenance)"


MENU_ORDER = (
    Modules.TABLEAU_DE_BORD,
    Modules.ETUDIANTS,
    Modules.INSCRIPTIONS,
    Modules.REFERENTIEL,
    Modules.FORMATEURS,
    Modules.NOTES,
    Modules.BULLETINS,
    Modules.EXAMENS,
    Modules.EMPLOI_DU_TEMPS,
    Modules.ABSENCES,
    Modules.ECOLAGE,
    Modules.PAIE,
    Modules.STATISTIQUES,
    Modules.RAPPORTS,
    Modules.ADMINISTRATION,
    Modules.UTILISATEURS,
    Modules.TABLES,
)

_DROITS = {
    Profils.ADMINISTRATEUR: None,  # tout
    Profils.DIRECTION: {
        Modules.ETUDIANTS, Modules.INSCRIPTIONS, Modules.REFERENTIEL, Modules.FORMATEURS,
        Modules.NOTES, Modules.BULLETINS, Modules.EXAMENS, Modules.EMPLOI_DU_TEMPS,
        Modules.ABSENCES, Modules.ECOLAGE, Modules.PAIE, Modules.STATISTIQUES,
        Modules.RAPPORTS,
    },
    Profils.SCOLARITE_LIBELLE: {
        Modules.ETUDIANTS, Modules.INSCRIPTIONS, Modules.REFERENTIEL, Modules.FORMATEURS,
        Modules.NOTES, Modules.BULLETINS, Modules.EXAMENS, Modules.EMPLOI_DU_TEMPS,
        Modules.ABSENCES, Modules.STATISTIQUES, Modules.RAPPORTS,
    },
    Profils.COMPTABILITE: {
        Modules.ECOLAGE, Modules.PAIE, Modules.STATISTIQUES, Modules.RAPPORTS,
    },
    Profils.ENSEIGNANT: {
        Modules.NOTES, Modules.BULLETINS, Modules.EXAMENS, Modules.EMPLOI_DU_TEMPS,
        Modules.ABSENCES, Modules.FORMATEURS,
    },
}

PROFILS_CONNUS = (
    Profils.ADMINISTRATEUR,
    Profils.DIRECTION,
    Profils.SCOLARITE_LIBELLE,
    Profils.COMPTABILITE,
    Profils.ENSEIGNANT,
)


def normalize_profil(profil: Optional[str]) -> str:
    """Ramène les libellés de la base aux profils normalisés."""
    p = (profil or "").strip()
    low = p.lower()
    if low in ("admin", "administrateur"):
        return Profils.ADMINISTRATEUR
    if low in ("direction",):
        return Profils.DIRECTION
    if low in ("scolarite", "scolarité"):
        return Profils.SCOLARITE_LIBELLE
    if low in ("finance", "comptabilite", "comptabilité", "caisse"):
        return Profils.COMPTABILITE
    if low in ("enseignant", "formateur"):
        return Profils.ENSEIGNANT
    return p if p.strip() else "Utilisateur"


def can_access(profil: Optional[str], module: str) -> bool:
    """Le profil peut-il ouvrir ce module ? (tableau de bord ouvert à tous)"""
    if module == Modules.TABLEAU_DE_BORD:
        return True
    role = normalize_profil(profil)
    droits = _DROITS.get(role)
    if droits is None and role == Profils.ADMINISTRATEUR:
        return True
    if droits is None:
        return False
    return module in droits
